src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageDrivenChannelAdapter.java

```diff
@@ -152,28 +195,43 @@ public void setFilterInRetry(boolean filterInRetry) {
 	protected void onInit() {
 		super.onInit();
 
-		AcknowledgingMessageListener<K, V> listener = this.listener;
+		if (this.mode.equals(ListenerMode.record)) {
+			AcknowledgingMessageListener<K, V> listener = this.recordListener;
 
-		boolean filterInRetry = this.filterInRetry && this.retryTemplate != null && this.recordFilterStrategy != null;
+			boolean filterInRetry = this.filterInRetry && this.retryTemplate != null && this.recordFilterStrategy != null;
 
-		if (filterInRetry) {
-			listener = new FilteringAcknowledgingMessageListenerAdapter<>(listener, this.recordFilterStrategy,
-					this.ackDiscarded);
-			listener = new RetryingAcknowledgingMessageListenerAdapter<>(listener, this.retryTemplate,
-					this.recoveryCallback);
+			if (filterInRetry) {
+				if (this.recordFilterStrategy != null) {
+					listener = new FilteringAcknowledgingMessageListenerAdapter<>(listener, this.recordFilterStrategy,
+						this.ackDiscarded);
+				}
+				if (this.retryTemplate != null) {
+					listener = new RetryingAcknowledgingMessageListenerAdapter<>(listener, this.retryTemplate,
+							this.recoveryCallback);
+				}
+			}
+			else {
+				if (this.retryTemplate != null) {
+					listener = new RetryingAcknowledgingMessageListenerAdapter<>(listener, this.retryTemplate,
+							this.recoveryCallback);
+				}
+				if (this.recordFilterStrategy != null) {
+					listener = new FilteringAcknowledgingMessageListenerAdapter<>(listener, this.recordFilterStrategy,
+							this.ackDiscarded);
+				}
+			}
+			this.messageListenerContainer.getContainerProperties().setMessageListener(listener);
 		}
 		else {
-			if (this.retryTemplate != null) {
-				listener = new RetryingAcknowledgingMessageListenerAdapter<>(listener, this.retryTemplate,
-						this.recoveryCallback);
-			}
+			BatchAcknowledgingMessageListener<K, V> listener = this.batchListener;
+
+			Assert.state(this.retryTemplate == null, "Retry is not supported with mode=batch");
```

Comment:
I think we can do this assertion directly in the `setRetryTemplate()`, but OTOH it doesn't hurt just to ignore it as is.